### PR TITLE
Key Agreement

### PR DESCRIPTION
**Pull request by Steppschuh**
The goal of this PR is to provide an easy way for two parties to agree on a secret symmetric key if they have previously exchanged their public keys.

### CODE REVIEW

File: rxkeystore/src/androidTest/java/com/nexenio/rxkeystore/provider/BaseCryptoProviderTest.java [hunk 1]
```diff
@@ -25,16 +29,34 @@
     protected RxKeyStore keyStore;
     protected RxCryptoProvider cryptoProvider;
 
-    protected void setUp() {
+    protected void setUpBeforeEachTest() {
```
Comment (marvinmirtschin) [inline rxkeystore/src/androidTest/java/com/nexenio/rxkeystore/provider/BaseCryptoProviderTest.java]: I guess we should add `@CallSuper`.

File: rxkeystore/src/androidTest/java/com/nexenio/rxkeystore/provider/asymmetric/rsa/RxRSACryptoProviderTest.java [hunk 2]
```diff
@@ -5,22 +5,44 @@
...
+    @Ignore("DHKeyAgreement requires DHPrivateKey")
+    @Test
+    @Override
+    public void generateSecretKey_matchingKeyPairs_sameSecretKey() {
+        super.generateSecretKey_matchingKeyPairs_sameSecretKey();
+    }
+
+    @Test
+    public void encrypt_invalidDataLength_emitsError() {
+        byte[] unencryptedBytes = LOREM_IPSUM_LONG.getBytes(StandardCharsets.UTF_8);
```
Comment (marvinmirtschin) [inline rxkeystore/src/androidTest/java/com/nexenio/rxkeystore/provider/asymmetric/rsa/RxRSACryptoProviderTest.java]: In another test you changed `LOREM_IPSUM_LONG.getBytes(StandardCharsets.UTF_8)` to `new byte[]{}`.

File: rxkeystore/src/main/java/com/nexenio/rxkeystore/provider/asymmetric/BaseAsymmetricCryptoProvider.java [hunk 3]
```diff
@@ -169,10 +199,130 @@ public Completable verify(@NonNull byte[] data, @NonNull byte[] signature, @NonN
         return Maybe.zip(getPublicKeyIfAvailable(alias), getPrivateKeyIfAvailable(alias), KeyPair::new);
     }
 
+    @Override
+    public Completable setKeyPair(@NonNull String alias, @NonNull KeyPair keyPair) {
+        return createSelfSignedCertificate(keyPair)
+                .flatMapCompletable(certificate -> Completable.mergeArray(
+                        //setCertificate(alias, new KeyStore.TrustedCertificateEntry(certificate)),
```
Comment (marvinmirtschin) [inline rxkeystore/src/main/java/com/nexenio/rxkeystore/provider/asymmetric/BaseAsymmetricCryptoProvider.java]: Out-commented code.

File: rxkeystore/src/androidTest/java/com/nexenio/rxkeystore/provider/asymmetric/rsa/RxRSACryptoProviderTest.java (hunk 2 above)
Comment (Steppschuh) [reply rxkeystore/src/androidTest/java/com/nexenio/rxkeystore/provider/asymmetric/rsa/RxRSACryptoProviderTest.java]: This is actually a different test (also different name), that tests input data that is longer than the block size.

**Pull request merged**